solr/core/src/java/org/apache/solr/update/TransactionLog.java

```diff
@@ -190,14 +205,14 @@ public boolean writePrimitive(Object val) throws IOException {
 
         long start = Files.size(tlog);
         channel = FileChannel.open(tlog, StandardOpenOption.READ, StandardOpenOption.WRITE);
-        os = Channels.newOutputStream(channel);
-        fos = new FastOutputStream(os, new byte[65536], 0);
-
         if (start > 0) {
           readHeader(null);
+        }
```

Comment:
Change the order: first read the header, and then open the output stream.
This does not change this TransactionLog behavior, but it allows an extending class to read some metadata in the header and use it to open the output stream.